util, ParseClientIP: support compressed ipv6 addresses

Acedus opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #
Previously ParseClientIP would use regex to match addresses which was too strict and could only correctly match complete ipv4 and ipv6 addresses.

This commit changes the ParseClientIP function to use Go's net package to correctly parse all valid ip forms.

## Is there anything that requires special attention ##

Do you have any questions?
No

Is the change backward compatible?
Yes

Are there concerns around backward compatibility?
No

Provide any external context for the change, if any.

For example:

* Kubernetes links that explain why the change is required
* CSI spec related changes/catch-up that necessitates this patch
* golang related practices that necessitates this change

## Related issues ##

Mention any github issues relevant to this PR. Adding below line
will help to auto close the issue once the PR is merged.

Fixes: #issue_number

## Future concerns ##

List items that are not part of the PR and do not impact it's
functionality, but are work items that can be taken up subsequently.

**Checklist:**

* [ ] **Commit Message Formatting**: Commit titles and messages follow
  guidelines in the [developer
  guide](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages).
* [ ] Reviewed the developer guide on [Submitting a Pull
  Request](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#development-workflow)
* [ ] [Pending release
  notes](https://github.com/ceph/ceph-csi/blob/devel/PendingReleaseNotes.md)
  updated with breaking and/or notable changes for the next major release.
* [ ] Documentation has been updated, if necessary.
* [ ] Unit tests have been added, if necessary.
* [ ] Integration tests have been added, if necessary.

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

Acedus commented:
fyi @akalenyu

akalenyu commented on internal/util/util_test.go:
```diff
@@ -300,10 +300,16 @@ func TestParseClientIP(t *testing.T) {
 		{
 			name:    "IPv6 address",
-			addr:    "2001:0db8:85a3:0000:0000:8a2e:0370:7334:0/2686266785",
+			addr:    "[2001:0db8:85a3:0000:0000:8a2e:0370:7334]:0/2686266785",
```
can the new impl. digest both? I wonder why the unit test was doing bracketless

Acedus replied:
New impl can't digest bracketless, but this should be the correct and mandatory notation as defined by [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.2) for ipv6 addresses with a specified port, as is the case here.

akalenyu replied:
I see, should be okay, any backwards compatibility concern would be hopefully noted by maintainers

nixpanic replied:
I don't think `2001:0db8:85a3:0000:0000:8a2e:0370:7334:0` is a valid formatted IPv6 address, as the last `:0` refers to a port (or wildcard in case of 0 with the Ceph session).

The format `[2001:0db8:85a3:0000:0000:8a2e:0370:7334]:0` looks more correct to me.

akalenyu reviewed (approved):
/cc @nixpanic 

seems like this is a bug in 3.15

Acedus commented:
[Change](https://github.com/ceph/ceph-csi/compare/60995439c4d6610a1c9044cb7e895e4a781252bf..da096eccdfc6d3f6e00aef3c7c1460c359c5ff9e) - Fix lint issues.

iPraveenParihar commented:
Hey @Acedus, thanks for the PR.
`fix(util)` is not an valid component in commit title (follow - [guide](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages))

You could just use `util: <subject>` format

Rakshith-R commented:
https://github.com/ceph/ceph-csi/actions/runs/17289568981/job/49168477608?pr=5547
@Acedus go-test is still failing
```
=== CONT  TestFetchIP/client.4305_2001:0db8:85a3:0000:0000:8a2e:0370:7334:0/422650892
    fencing_test.go:95: expected error false but got failed to extract IP address, incorrect format: address client.4305 2001:0db8:85a3:0000:0000:8a2e:0370:7334:0/422650892: too many colons in address
    fencing_test.go:99: expected IP 2001:db8:85a3::8a2e:370:7334 but got 
--- FAIL: TestFetchIP (0.00s)
    --- FAIL: TestFetchIP/client.4305_172.21.9.34:0/422650892 (0.00s)
    --- PASS: TestFetchIP/#00 (0.00s)
    --- FAIL: TestFetchIP/client.24152_v1:100.64.0.7:0/3658550259 (0.00s)
    --- FAIL: TestFetchIP/client.4305_2001:0db8:85a3:0000:0000:8a2e:0370:7334:0/422650892 (0.00s)
FAIL
```

nixpanic commented on internal/util/util.go:
```diff
@@ -302,29 +301,17 @@ func GetVolumeContext(parameters map[string]string) map[string]string {
-		if ip != nil {
-			return ip.String(), nil
-		}
+	host, _, err := net.SplitHostPort(addr)
```
Before doing the `SplitHostPort`, make sure to take only the address+port. The `addr` string may contain an extra `/012345` nonce for identifying a unique session.

Doing it this way, is much nicer than a hugely complex regular expression. Thanks!

Acedus replied:
Yeah, there are a few things I forgot to consider like the version prefix among other things, will fix.

Acedus committed:
util: allow ParseClientIP to handle compressed ipv6 addresses

Previously ParseClientIP would use regex to match addresses which was
too strict and could only correctly match complete ipv4 and ipv6
addreses.

This commit changes the ParseClientIP function to use Go's net package
to correctly parse all valid ip forms.

Signed-off-by: Adi Aloni <aaloni@redhat.com>

Acedus commented:
[Change](https://github.com/ceph/ceph-csi/compare/1fdbff38c85f97e675e13217ee62755abfbbebb6..e66e6dd47ef5459fab2165325e35aa5cdbc65eb5) - Address comments, take care to iterate on valid host pairs and omit version prefix before parse attempts.

Rakshith-R commented:
@mergifyio queue

Rakshith-R commented:
@mergifyio rebase

iPraveenParihar commented:
@Mergifyio queue

Rakshith-R commented:
@mergifyio requeue
